libnetwork/sandbox_dns_unix.go

```diff
@@ -48,7 +49,8 @@ func (sb *Sandbox) startResolver(restore bool) {
 		// have a gateway. So, if the Sandbox is only connected to an 'internal' network,
 		// it will not forward DNS requests to external resolvers. The resolver's
 		// proxyDNS setting is then updated as network Endpoints are added/removed.
-		sb.resolver = NewResolver(resolverIPSandbox, sb.getGatewayEndpoint() != nil, sb)
+		proxyDNS := !sb.config.dnsNoProxy && sb.getGatewayEndpoint() != nil
+		sb.resolver = NewResolver(resolverIPSandbox, proxyDNS, sb)
```

Comment:
What's stopping `sb.resolver.SetForwardingPolicy()` calls (#47538) from overriding the `dnsNoProxy` config when joining and leaving endpoints to the sandbox?

Reply:
Oh! Thank you, that was silly of me.

I can only think of ugly fixes. `Resolver` could have an `allowProxyDNS` to use as ...

```
func (r *Resolver) serveDNS(w dns.ResponseWriter, query *dns.Msg) {
	...
	if r.allowProxyDNS && r.proxyDNS.Load() {
```

Making it an enum, disable/on/off, would be nicer but then `SetForwardingPolicy()` and the atomic.Bool get messier.

To initialise it, could have an extra constructor ...

```
	if sb.config.dnsNoProxy {
		sb.resolver = NewResolverNoProxy(resolverIPSandbox, sb)
	} else {
		sb.resolver = NewResolver(resolverIPSandbox, sb.getGatewayEndpoint() != nil, sb)
	}
```

Or, initialise it to `proxyDNS` and requiring a `SetForwardingPolicy(false)` to disable it if the initial networks are all internal ...

```
	sb.resolver = NewResolver(resolverIPSandbox, !sb.config.dnsNoProxy, sb)
	sb.resolver.SetForwardingPolicy(sb.getGatewayEndpoint() != nil)
```

(Or an extra flag in the existing constructor, or enum instead of two flags - but that's a bit awkward.)

Can you think of something pretty?

Reply:
Instead of the Endpoint reaching into `sb.resolver`, have the sandbox be in charge of calling `sb.resolver.SetForwardingPolicy` somehow so that it can gate enabling of forwarding on `sb.config.dnsNoProxy`. While it'd be really elegant if that could be piggybacked off an existing sandbox method called by `func (*Endpoint) sbJoin`, I didn't spot an obvious choice that wouldn't be problematic in some way. `sb.updateDNS` looks like a promising candidate, but the call would have to be moved to after the `sb.addEndpoint(ep)` call, which risks unintended side effects. Go with the pragmatic solution of writing a new sandbox method which `func (*Endpoint) sbJoin` and `func (*Endpoint) sbLeave` call instead of `sb.resolver.SetForwardingPolicy`?
```go
func (sb *Sandbox) RefreshResolverPolicy() {
	sb.resolver.SetForwardingPolicy(!sb.config.dnsNoProxy && sb.getGatewayEndpoint() != nil)
}
```

Reply:
Perfect - thank you.

(I'm about to publish a PR that removes `sb.UpdateDNS`, so I'm glad it wasn't the right place!)